drivers/espi/espi_realtek_rts5912.c

```diff
@@ -0,0 +1,236 @@
+		break;
+	case RTS5912_ESPI_MAX_FREQ_25:
+		gen_conf |= 1UL << ESPI_ESPICFG_MXFREQSUP_Pos;
+		break;
+	case RTS5912_ESPI_MAX_FREQ_33:
+		gen_conf |= 2UL << ESPI_ESPICFG_MXFREQSUP_Pos;
+		break;
+	case RTS5912_ESPI_MAX_FREQ_50:
+		gen_conf |= 3UL << ESPI_ESPICFG_MXFREQSUP_Pos;
+		break;
+	case RTS5912_ESPI_MAX_FREQ_66:
+		gen_conf |= 4UL << ESPI_ESPICFG_MXFREQSUP_Pos;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	/* I/O Mode Supported */
+	io_mode = cfg->io_caps >> 1;
+
+	if (io_mode > 3) {
+		return -EINVAL;
+	}
+	gen_conf |= io_mode << ESPI_ESPICFG_IOSUP_Pos;
+
+	/* Channel Supported */
+	if (cfg->channel_caps & ESPI_CHANNEL_PERIPHERAL) {
+		gen_conf |= BIT(0) << ESPI_ESPICFG_CHSUP_Pos;
+	}
+
+	if (cfg->channel_caps & ESPI_CHANNEL_VWIRE) {
+		gen_conf |= BIT(1) << ESPI_ESPICFG_CHSUP_Pos;
+	}
+
+	if (cfg->channel_caps & ESPI_CHANNEL_OOB) {
+		gen_conf |= BIT(2) << ESPI_ESPICFG_CHSUP_Pos;
+	}
+
+	if (cfg->channel_caps & ESPI_CHANNEL_FLASH) {
+		gen_conf |= BIT(3) << ESPI_ESPICFG_CHSUP_Pos;
+	}
+
+	espi_reg->ESPICFG = gen_conf;
+	data->config_data = espi_reg->ESPICFG;
+
+	return 0;
+}
+
+static bool espi_rts5912_channel_ready(const struct device *dev, enum espi_channel ch)
+{
+	const struct espi_rts5912_config *const espi_config = dev->config;
+	volatile struct espi_reg *const espi_reg = espi_config->espi_reg;
+
+	switch (ch) {
+	case ESPI_CHANNEL_PERIPHERAL:
+		return espi_reg->EPCFG & ESPI_EPCFG_CHEN ? true : false;
+	case ESPI_CHANNEL_VWIRE:
+		return espi_reg->EVCFG & ESPI_EVCFG_CHEN ? true : false;
+	case ESPI_CHANNEL_OOB:
+		return espi_reg->EOCFG & ESPI_EOCFG_CHEN ? true : false;
+	case ESPI_CHANNEL_FLASH:
+		return espi_reg->EFCONF & ESPI_EFCONF_CHEN ? true : false;
+	default:
+		return false;
+	}
+}
+
+static int espi_rts5912_manage_callback(const struct device *dev, struct espi_callback *callback,
+					bool set)
+{
+	struct espi_rts5912_data *data = (struct espi_rts5912_data *)dev->data;
+
+	return espi_manage_callback(&data->callbacks, callback, set);
+}
+
+static DEVICE_API(espi, espi_rts5912_driver_api) = {
+	.config = espi_rts5912_configure,
+	.get_channel_status = espi_rts5912_channel_ready,
+	.manage_callback = espi_rts5912_manage_callback,
+};
+
+static void espi_rst_isr(const struct device *dev)
+{
+	const struct espi_rts5912_config *const espi_config = dev->config;
+	struct espi_rts5912_data *data = (struct espi_rts5912_data *)dev->data;
+
+	struct espi_event evt = {.evt_type = ESPI_BUS_RESET, .evt_details = 0, .evt_data = 0};
+
+	volatile struct espi_reg *const espi_reg = espi_config->espi_reg;
+	uint32_t status = espi_reg->ERSTCFG;
+
+	espi_reg->ERSTCFG |= ESPI_ERSTCFG_RSTSTS;
+	espi_reg->ERSTCFG ^= ESPI_ERSTCFG_RSTPOL;
+
+	if (status & ESPI_ERSTCFG_RSTSTS) {
+		if (status & ESPI_ERSTCFG_RSTPOL) {
+			/* rst pin high go low trigger interrupt */
+			evt.evt_data = 0;
+		} else {
+			/* rst pin low go high trigger interrupt */
+			evt.evt_data = 1;
+		}
+		espi_send_callbacks(&data->callbacks, dev, evt);
```

Comment:
It looks like this shouldn't compile?
```suggestion
		espi_send_callbacks(&data->callbacks, dev, &evt);
```

Reply:
static inline void espi_send_callbacks(sys_slist_t *list,
				       const struct device *dev,
				       struct espi_event evt)
The function is pass evt by value

Reply:
My mistake - you're correct.